hashicorp/consul
# xDS Load Balancing

### Description
Prior to #13244, connect proxies and gateways could only be configured by an xDS session served by the local client agent. 

In an upcoming release, it will be possible to deploy a Consul service mesh without client agents. In this model, xDS sessions will be handled by the servers themselves, which necessitates load-balancing to prevent a single server from receiving a disproportionate amount of load and becoming overwhelmed.

This PR introduces a simple form of load-balancing where Consul will attempt to achieve an even spread of load (xDS sessions) between all healthy servers. It does so by implementing a concurrent session limiter (`limiter.SessionLimiter`) and adjusting the limit according to autopilot state and proxy service registrations in the catalog.

If a server is already over capacity (i.e. the session limit is lowered), Consul will begin draining sessions to rebalance the load. This will result in the client receiving a `RESOURCE_EXHAUSTED` status code. It is the client's responsibility to observe this response and reconnect to a different server. Users of the gRPC client connection brokered by the [consul-server-connection-manager](https://github.com/hashicorp/consul-server-connection-manager) library will get this for free.

The rate at which Consul will drain sessions to rebalance load is scaled dynamically based on the number of proxies in the catalog:

| Number of proxies | Drain rate |
| - | - |
|0-512|1 session per second|
|513-2815|linearly scaled by 1/second for every additional 256 proxies|
|2816+|10 sessions per second|

### Testing Steps
- Start a Consul dev server (`consul agent -dev`)
- Register 10 proxy services using the HTTP API
- Start Envoy proxies for them (tip: folks on the Consul team can use the `consul-dataplane` binary for this!)
- Join another server to the cluster (`consul agent -dev -retry-join "127.0.0.1:8301" -server-port 7300 -serf-wan-port -1 -serf-lan-port 7301 -dns-port -1 -http-port -1 -grpc-port -1 -node "other-server"`)
- You should see ~4 proxies receive `RESOURCE_EXHAUSTED` errors as the server tries to rebalance load.
